zmap: init at 2.1.1

Ma27 opened the pull request:
<!-- Nixpkgs has a lot of new incoming Pull Requests, but not enough people to review this constant stream. Even if you aren't a committer, we would appreciate reviews of other PRs, especially simple ones like package updates. Just testing the relevant package/service and leaving a comment saying what you tested, how you tested it and whether it worked would be great. List of open PRs: <https://github.com/NixOS/nixpkgs/pulls>, for more about reviewing contributions: <https://hydra.nixos.org/job/nixpkgs/trunk/manual/latest/download/1/nixpkgs/manual.html#sec-reviewing-contributions>. Reviewing isn't mandatory, but it would help out a lot and reduce the average time-to-merge for all of us. Thanks a lot if you do! -->
###### Motivation for this change
zmap[1] is a fast network scanner for the IPv4 address space. This is
the main package of the ZMap projects, there are further one that will
be packaged soon.

[1]: https://zmap.io/

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nix-review --run "nix-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Assured whether relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

Ma27 committed:
zmap: init at 2.1.1

zmap[1] is a fast network scanner for the IPv4 address space. This is
the main package of the ZMap projects, there are further one that will
be packaged soon.

[1] https://zmap.io/

etu reviewed (approved):
Built locally and it seems to work :-)

volth commented on pkgs/tools/security/zmap/0001-The-byacc-binary-in-nixpkgs.byacc-is-named-yacc.patch:
```diff
@@ -0,0 +1,26 @@
+ else()
+ add_custom_command(OUTPUT parser.c
+-    COMMAND byacc -d -o "${CMAKE_CURRENT_BINARY_DIR}/parser.c" "${CMAKE_CURRENT_SOURCE_DIR}/parser.y"
++    COMMAND yacc -d -o "${CMAKE_CURRENT_BINARY_DIR}/parser.c" "${CMAKE_CURRENT_SOURCE_DIR}/parser.y"
```
Isn't it an overkill to create a .patch file just for ```s/byacc/yacc/``` ?
Maybe just
```postPatch = "substituteInPlace CMakeLists.txt --replace byacc yacc"```;

Ma27 replied:
I originally added the patch as there was some more stuff to do that I managed to fix without the patch. I don't bother too much about small patches which is why I kept it.
But if everybody else agress, I can fix that :)

worldofpeace replied:
I'm strong supporter here of a patch because of what kind of failure it would give as compared to `substituteInPlace` where it may build without the substitution. A patch will just fail to apply.

volth replied:
Ok, but

* patch applications often fail because of changes in the patched files which have been captured into ```.patch```-file being irrelevant to the intended fix.
Here we have to change a single word but captured ~15 words to track them
* if one more ```byacc``` appeared in the next version, patch won't fix it

worldofpeace replied:
Arch seems to do `--program-transform-name='s,^,b,'` as a `configureFlag`, Seems we should just do that since the executable probably should be called `byacc` anyway. (it is the packages name)
See also what fedora does https://src.fedoraproject.org/rpms/byacc/blob/master/f/byacc.spec#_37

##### train of thought is doing a substitution at a lower level :smile: 

@Ma27 does this sound correct to you?

Ma27 replied:
That's indeed a good idea. I'll have a look at this :)

Ma27 replied:
It seems as this is something that only works with `./configure` scripts generated by `m4`, but not by CMake. I guess it's easier to use `substituteInPlace` then.

worldofpeace commented on pkgs/tools/security/zmap/default.nix:
```diff
@@ -0,0 +1,32 @@
+  meta = with stdenv.lib; {
+    homepage = https://zmap.io/;
+    license = licenses.asl20;
+    description = "ZMap is a fast single packet network scanner designed for Internet-wide network surveys.";
```
```suggestion
    description = "Fast single packet network scanner designed for Internet-wide network surveys";
```

worldofpeace commented on pkgs/tools/security/zmap/default.nix:
```diff
@@ -0,0 +1,32 @@
+
+  patches = [ ./0001-The-byacc-binary-in-nixpkgs.byacc-is-named-yacc.patch ];
+
+  cmakeFlags = [ "-DRESPECT_INSTALL_PREFIX_CONFIG=ON" ];
```
So this will install the config within our prefix. Will it still read it at `/etc`? This is a common problem.

Ma27 replied:
It seems as it still reads from `/etc/zmap/` by default, but it's changeable using `-C`.

> This is a common problem.


I know. This will work fine when installing the package in `environment.systemPackages`, but will cause problems if not I guess. What would you suggest here?

worldofpeace commented on pkgs/tools/security/zmap/default.nix:
```diff
@@ -0,0 +1,32 @@
+
+  outputs = [ "out" "man" ];
+
+  meta = with stdenv.lib; {
```
`platforms` is not set. It appears to support a wider variety.

Ma27 replied:
Unfortunately I don't have the resources and experience to fix broken darwin builds. I actually hoped that darwin would've worked out of the box and wanted to wait for ofborg's results.

My suggestion is that we restrict the build to linux platforms for now and let the issue fix someone with more experience here, if the demanded :)

worldofpeace replied:
Oh no, I don't expect you to do anything outside your means.

I think you can still set platforms but mark it as broken for darwin.

worldofpeace reviewed (changes requested):
See above comments. Also `darwin` build is failing and should be fixed.

BenBals commented:
##### Reviewed points

- [x] package path fits guidelines
- [x] package name fits guidelines
- [x] package version fits guidelines
- [x] package build on Linux nixos-beb 4.19.37 #1-NixOS SMP Sat Apr 27 07:36:41 UTC 2019 x86_64 GNU/Linux
- [x] executables tested on Linux nixos-beb 4.19.37 #1-NixOS SMP Sat Apr 27 07:36:41 UTC 2019 x86_64 GNU/Linux
- [x] `meta.description` is set and fits guidelines (With the proposed change by @worldofpeace)
- [x] `meta.license` fits upstream license
- [ ] `meta.platforms` is set
- [x] `meta.maintainers` is set
- [x] source is fetched using the appropriate function

##### Possible improvements

##### Comments

Ma27 commented:
@worldofpeace as far as I see I addressed all of the review comments except the one with the config file in`/etc`. I'm not sure that's the best solution there, do you have a suggestion?:)

worldofpeace commented on pkgs/tools/security/zmap/default.nix:
```diff
@@ -0,0 +1,36 @@
+  meta = with stdenv.lib; {
+    homepage = https://zmap.io/;
+    license = licenses.asl20;
+    description = "Fast single packet network scanner designed for Internet-wide network surveys.";
```
```suggestion
    description = "Fast single packet network scanner designed for Internet-wide network surveys";
```

No periods at the end of `description`

worldofpeace commented:
> @worldofpeace as far as I see I addressed all of the review comments except the one with the config file in`/etc`. I'm not sure that's the best solution there, do you have a suggestion?:)

You've addressed all the comments except  https://github.com/NixOS/nixpkgs/pull/61048#discussion_r282561240 https://github.com/NixOS/nixpkgs/pull/61048#discussion_r282652110

I'm think platforms should represent what's supported upstream, `unix` I think, but we can mark it as broken on darwin since we won't be working for that.

And for the config file situation, it depends on how it's intended to be used in NixOS.
For example, it default reading the config from `/etc` and not the bundled one in the prefix is useful if there is a nixos module which just adds it to `environment.systemPackages` and uses `environment.etc` to link the config from within its output.

worldofpeace commented:
Also doing
```diff
diff --git a/pkgs/development/tools/parsing/byacc/default.nix b/pkgs/development/tools/parsing/byacc/default.nix
index ba1b8f27e7c..fe8b903180e 100644
--- a/pkgs/development/tools/parsing/byacc/default.nix
+++ b/pkgs/development/tools/parsing/byacc/default.nix
@@ -14,6 +14,10 @@ stdenv.mkDerivation rec {
 
   doCheck = true;
 
+  configureFlags = [
+    "--program-transform-name='s,^,b,'"
+  ];
+
   meta = with stdenv.lib; {
     description = "Berkeley YACC";
     homepage = https://invisible-island.net/byacc/byacc.html;
diff --git a/pkgs/tools/security/zmap/default.nix b/pkgs/tools/security/zmap/default.nix
index 9e509ed3d0e..03395cb492a 100644
--- a/pkgs/tools/security/zmap/default.nix
+++ b/pkgs/tools/security/zmap/default.nix
@@ -13,11 +13,6 @@ stdenv.mkDerivation rec {
     sha256 = "0yaahaiawkjk020hvsb8pndbrk8k10wxkfba1irp12a4sj6rywcs";
   };
 
-  postPatch = ''
-    substituteInPlace src/CMakeLists.txt \
-      --replace "byacc" "yacc"
-  '';
-
   cmakeFlags = [ "-DRESPECT_INSTALL_PREFIX_CONFIG=ON" ];
   dontUseCmakeBuildDir = true;
 
```

fixed the issue with `byacc` not being found for me...

##### not sure if there should be symlinks for backwards compat

Ma27 committed:
byacc: resulting binary should be named `byacc`

Ma27 commented:
> You've addressed all the comments except #61048 (comment) #61048 (comment)

Yeah I've pushed the `meta.platform` fix before I've seen your comment about marking it broken :)

> fixed the issue with byacc not being found for me...

Ahh, I'm afraid I misunderstood the idea behind `--program-transform-name`, sorry!

>And for the config file situation, it depends on how it's intended to be used in NixOS.
For example, it default reading the config from /etc and not the bundled one in the prefix is useful if there is a nixos module which just adds it to environment.systemPackages and uses environment.etc to link the config from within its output.

>  For example, it default reading the config from /etc and not the bundled one in the prefix is useful if there is a nixos module which just adds it to environment.systemPackages and uses environment.etc to link the config from within its output.

:+1:

Ma27 committed:
nixos/zmap: init module

The module installs `zmap` globally and links the config files to
`/etc/zmap`, the default location of config files for zmap.

The package provides pretty much a sensitive default, custom configs can
be created like this:

```
{ lib, ... }:
{
  environment.etc."zmap/blacklist.conf" = lib.mkForce {
    text = ''
      # custom zmap blacklist
      0.0.0.0/0
    '';
  };
}
```

worldofpeace reviewed (approved):
I think It's perfect @Ma27 :smile:

BenBals commented:
I am really happy to see this merged.
